File: Project2Files/UDPclient.c
#include <arpa/inet.h>
#include <netdb.h>
#include "UDPclient.h"


/*
 * UDPclient.c
 * Systems and Networks II
 * Project 2
 *
 */

//Constants
const int BUFFERSIZE = 256;



/*
 * Creates a datagram socket and connects to a server.
 *
 * serverName - the ip address or hostname of the server given as a string
 * serverPort - the port number of the server
 * dest       - the server's address information; the structure will be initialized with information
 *              on the server (like port, address, and family) in this function call
 *
 * return value - the socket identifier or a negative number indicating the error if a connection could not be established
 */
int createSocket(char * serverName, int port, struct sockaddr_in * dest)
{
    /*~~~~~~~~~~~~~~~~~~~~~Local vars~~~~~~~~~~~~~~~~~~~~~*/
    int socketFD;
    struct hostent *hostptr;
    struct in_addr ipAddress;
    /*~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/


    if( (socketFD = socket(AF_INET, SOCK_DGRAM, 0) ) < 0)
    {
        printf("Socket creation failed\n");
        return -1;
    }
    if( (hostptr = gethostbyname(serverName) ) == NULL)
    {
        perror("gethostbyname() failed, exit\n");
        return -1;
    }

    int i=0;
    printf("\nIP: " );
    while(hostptr->h_addr_list[i] != 0)
    {
        ipAddress.s_addr = *(u_long*)hostptr->h_addr_list[i++];
        printf("%s\n", inet_ntoa(ipAddress));
    }

    memset((void*)dest, 0, sizeof(struct sockaddr_in));    /* zero the struct */
    dest->sin_family = AF_INET;
    memcpy( (void *)&dest->sin_addr, (void *)hostptr->h_addr, hostptr->h_length);
    dest->sin_port = htons( (u_short)port );        /* set destination port number */
    printf("port: %d\n", htons(dest->sin_port));
    return socketFD;
}
/*
 * Sends a request for service to the server. This is an asynchronous call to the server,
 * so do not wait for a reply in this function.
 *
 * sockFD  - the socket identifier
 * request - the request to be sent encoded as a string
 * dest    - the server's address information
 *
 * return   - 0, if no error; otherwise, a negative number indicating the error
 */
int sendRequest(int socketFD, char * request, struct sockaddr_in * dest)
{
    socklen_t destSize = sizeof(struct sockaddr_in);
    int size = sendto(socketFD, request, strlen(request), 0, (struct sockaddr *) dest, destSize);
    return size;
}
/*
 * Receives the server's response formatted as an XML text string.
 *
 * sockfd    - the socket identifier
 * response  - the server's response as an XML formatted string to be filled in by this function into the specified string array
 *
 * return   - 0, if no error; otherwise, a negative number indicating the error
 */
int receiveResponse(int socketFD, char * response, int size)
{
    int length = 0;
    socklen_t bufSize = (socklen_t)size;
    length = recvfrom(socketFD, response, bufSize, 0, NULL, NULL);
    response[length] = '\0';
    return length;
}

/*
 * Prints the response to the screen in a formatted way.
 *
 * response - the server's response as an XML formatted string
 *
 */
void printResponse(char * response)
{
    printf("%s\n",response);
}

/*
 * Closes the specified socket
 *
 * sockFD - the ID of the socket to be closed
 *
 * return - 0, if no error; otherwise, a negative number indicating the error
 */
int closeSocket(int socketFD)
{
    close(socketFD);
    return 0;
}

File: Project2Files/UDPclient.java
/**
 * Created by brett on 2/5/15.
 */
import java.io.*;
import java.net.*;
import java.net.DatagramSocket;

/*
 * UDPclient.java
 * Systems and Networks II
 * Project 2
 *
 * This file describes the functions to be implemented by the UDPclient class
 * You may also implement any helper functions you deem necessary to complete the project.
 */

public class UDPclient
{
    private DatagramSocket _socket; // the socket for communication with a server
    DatagramPacket packet;
    byte[] buf;

    /**
     * Constructs a UDPclient object.
     */
    public UDPclient()
    {
    }

    /**
     * Creates a datagram socket and binds it to a free port.
     *
     * @return - 0 or a negative number describing an error code if the connection could not be established
     */
    public int createSocket()
    {
        try {
            _socket = new DatagramSocket();
        }
        catch(Exception e)
        {
            System.out.println("New socket failed");
            return -1;
        }
        return 0;
    }

    /**
     * Sends a request for service to the server. Do not wait for a reply in this function. This will be
     * an asynchronous call to the server.
     *
     * @param request - the request to be sent
     * @param hostAddr - the ip or hostname of the server
     * @param port - the port number of the server
     *
     * @return - 0, if no error; otherwise, a negative number indicating the error
     */
    public int sendRequest(String request, String hostAddr, int port)
    {
        buf = new byte[256];

        for(int i = 0; i < request.length(); i++) {
            buf[i] = (byte)request.charAt(i);
        }

        try{
            request = request + '\0';
            InetAddress address = InetAddress.getByName(hostAddr);
            packet = new DatagramPacket(buf, buf.length, address, port);
            _socket.send(packet);
            System.out.println("Sent: " + request);
        }
        catch(Exception e)
        {
            System.out.println("sendRequest failed");
            return -1;
        }
        return 0;
    }

    /**
     * Receives the server's response following a previously sent request.
     *
     * @return - the server's response or NULL if an error occurred
     */
    public String receiveResponse()
    {
        packet = new DatagramPacket(buf, buf.length);
        String response;

        try{
            _socket.receive(packet);
            response = new String(packet.getData(), 0, packet.getLength());
            System.out.print("Received: ");
        }
        catch (Exception e)
        {
            System.out.println("receiveResponse() failed");
            return null;
        }
        return response;
    }

    /*
    * Prints the response to the screen in a formatted way.
    *
    * response - the server's response as an XML formatted string
    *
    */
    public static void printResponse(String response)
    {
        System.out.println("Received " + response);
    }


    /*
     * Closes an open socket.
     *
    * @return - 0, if no error; otherwise, a negative number indicating the error
     */
    public int closeSocket()
    {
        try{
            _socket.close();
        }
        catch(Exception e)
        {
            System.out.println("closeSocket() failed");
            return -1;
        }
        return 0;
    }
}